google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java

```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
+        return this;
+      }
+
+      public Builder addContentEncodingCondition(V4ConditionType type, String contentEncoding) {
+        conditions.add(V4Condition.newV4Condition(type, "content-encoding", contentEncoding));
+        return this;
+      }
+
+      public Builder addContentLengthCondition(V4ConditionType type, int contentLength) {
+        conditions.add(V4Condition.newV4Condition(type, "content-length", "" + contentLength));
+        return this;
+      }
+
+      public Builder addContentTypeCondition(V4ConditionType type, String contentType) {
+        conditions.add(V4Condition.newV4Condition(type, "content-type", contentType));
+        return this;
+      }
+
+      public Builder addExpiresCondition(V4ConditionType type, long expires) {
+        conditions.add(V4Condition.newV4Condition(type, "expires", dateFormat.format(expires)));
+        return this;
+      }
+
+      public Builder addExpiresCondition(V4ConditionType type, String expires) {
+        conditions.add(V4Condition.newV4Condition(type, "expires", expires));
+        return this;
+      }
+
+      public Builder addKeyCondition(V4ConditionType type, String key) {
+        conditions.add(V4Condition.newV4Condition(type, "key", key));
+        return this;
+      }
+
+      public Builder addSuccessActionRedirectUrlCondition(
+          V4ConditionType type, String successActionRedirectUrl) {
+        conditions.add(
+            V4Condition.newV4Condition(type, "success_action_redirect", successActionRedirectUrl));
+        return this;
+      }
+
+      public Builder addSuccessActionStatusCondition(V4ConditionType type, int status) {
+        conditions.add(V4Condition.newV4Condition(type, "success_action_status", "" + status));
+        return this;
+      }
+
+      public Builder addCustomMetadataCondition(V4ConditionType type, String field, String value) {
+        conditions.add(V4Condition.newV4Condition(type, "x-goog-meta-" + field, value));
+        return this;
+      }
+
+      public Builder addContentLengthRange(int min, int max) {
+        conditions.add(
+            V4Condition.newV4Condition(V4ConditionType.CONTENT_LENGTH_RANGE, "" + min, "" + max));
+        return this;
+      }
+
+      public Builder addCustomCondition(V4ConditionType type, String field, String value) {
+        conditions.add(V4Condition.newV4Condition(type, field, value));
+        return this;
+      }
+    }
+  }
+
+  class V4PostPolicyDocument {
+    private String expiration;
+    private V4PostConditions conditions;
+
+    private V4PostPolicyDocument(String expiration, V4PostConditions conditions) {
+      this.expiration = expiration;
+      this.conditions = conditions;
+    }
+
+    public static V4PostPolicyDocument of(String expiration, V4PostConditions conditions) {
+      return new V4PostPolicyDocument(expiration, conditions);
+    }
+
+    public String toJsonObject() {
```

Comment:
I'm not sure about this one but consider toJson, not to JsonObject in keeping with return type